### PR TITLE
Data streams v2

### LINKED ISSUE #2938: Deriving `uniffi::Error` on an enum with a variant that contains a `message` field leads to a Kotlin error
This error type in Rust

```rust
#[derive(Debug, thiserror::Error, uniffi::Error)]
pub enum FfiError {
    #[error("Cancelled")]
    Cancelled,
    #[error("Generic error")]
    GenericError,
    #[error("StringError {0}")]
    StringError(String),
    #[error("ApiError url: {url}, status_code: {status_code:?}, message: {message}")]
    ApiError {
        url: String,
        status_code: Option<u16>,
        message: String,
    },
}
```

works well in Swift, but when generating code for Kotlin, the following type is generated:

```kotlin
sealed class FfiException: kotlin.Exception() {
    class Cancelled(
        ) : FfiException() {
        override val message
            get() = ""
    }
    
    class GenericException(
        ) : FfiException() {
        override val message
            get() = ""
    }
    
    class StringException(
        val v1: kotlin.String
        ) : FfiException() {
        override val message
            get() = "v1=${ v1 }"
    }
    
    class ApiException(
        val `url`: kotlin.String,         
        val `statusCode`: kotlin.UShort?, 
        
        val `message`: kotlin.String
        ) : FfiException() {
        override val message
            get() = "url=${ `url` }, statusCode=${ `statusCode` }, message=${ `message` }"
    }

    companion object ErrorHandler : UniffiRustCallStatusErrorHandler<FfiException> {
        override fun lift(error_buf: RustBuffer.ByValue): FfiException = FfiConverterTypeFfiError.lift(error_buf)
    }   
}
```

Which gives the following compiler errors:

```shell
> Task :app:compileDebugKotlin FAILED
.../ffi.kt:11905:13 Conflicting declarations:
val message: String
.../ffi.kt:11905:13 'message' hides member of supertype 'Throwable' and needs an 'override' modifier.
.../ffi.kt:11907:22 Conflicting declarations:
val message: String
.../ffi.kt:11908:79 Overload resolution ambiguity between candidates:
val message: String
val message: String
.../ffi.kt:11964:59 Overload resolution ambiguity between candidates:
val message: String
val message: String
.../ffi.kt:11988:48 Overload resolution ambiguity between candidates:
val message: String
val message: String
```

This is in part a "bug report" and in part just a question: Can't I use the identifier `message` in my error types? Obviously there is a conflict here, but the cause isn't obvious. Maybe `uniffi::Error` should warn / error if I use an incompatible field name?

Workaround: Rename my `message` field to `msg`.

### PR DESCRIPTION
**Pull request by 1egoman**
Initial port of data streams v2 to the android sdk. The corresponding swift change (which this is fairly heavily patterned off of ) can be found here: https://github.com/livekit/client-sdk-swift/pull/1075

Previously, the android sdk had its own data streams implementation. With [data streams v2](https://github.com/livekit/client-sdk-js/pull/1985), the rust implementation will both be quite a bit more stable and gain some new features that make it significantly more performant (single packet data streams and DEFLATE compression when it makes the payload smaller). This roughly _doubles_ data stream throughput in local testing.

So, port the android sdk to use the rust sdk data streams v2 implementation, and completely remove the pre-existing kotlin implementation. This is a substantial change which needs thorough testing.

### New behaviors worth being aware of
All of these are either data streams v2 related changes, or bug fixes.

#### 1. `compress` option
When sending a data stream, there is a new `compress` option. Just like how this works on web / rust, this defaults to `true`. If set to `false`, then compression will be disabled (useful if you know the data you are sending isn't compressible / you are doing your own compression, which is not uncommon in robotics use cases). The vast majority of users should leave this set to `true`.

#### 2. Max data stream size
In a [rust data streams v2 pull request review comment](https://github.com/livekit/rust-sdks/pull/1192#discussion_r3571761970), we decided that for security reasons it made sense to introduce a maximum data stream size as a DOS protection. This limit is by default `5gb` - any data stream that is larger will now read up until that point, and if the stream keeps going, a "payload too large" error will be raised on the stream and exposed to a user on the subsequent `.read()` call.

If a user is sending a large file, they can override this by setting a new `maxPayloadByteLength` option on the `room.connect` call:
```kotlin
room.connect(
    url = wsUrl,
    token = token,
    options = ConnectOptions(
        autoSubscribe = true
    ),
    roomOptions = RoomOptions(
        dynacast = false,
        dataStream = DataStreamOptions(maxPayloadByteLength = 1000)
    )
)
```

#### 3. Throwing new data stream errors types
The old kotlin specific data streams implementation wasn't quite as strict and didn't surface as many error cases to the caller which were encountered while reading the stream as the rust implementation now does. This needs some testing in some of these edge cases to make sure I didn't inadvertently make this backwards incompatible in a non-aceptable way.

## Adding demo to `sample-app`

To exercise these changes, I've added a new section to the `sample-app` example which renders a data stream testing interface. This has been heavily patterned off of the similar interface I added to rust-dev-client [here](https://github.com/livekit-examples/rust-dev-client/pull/18):

Screenshot 2026-08-07 at 11 07 38 AM

## Uniffi integration

In addition to the data streams v2 features, I've started integrating the uniffi kotlin bindings into this sdk based off of DL's [in progress branch](https://github.com/livekit/client-sdk-android/tree/dl/rust-testing). For the most part, this has gone fairly smoothly. I've kept everything wired up with maven local for the time being and will leave it as a cleanup item for an android expert (likely DL) to this working properly prior to merging.

As part of this, I have also fixed two issues in the `livekit-uniffi` kotlin bindgen:
1. Any uniffi objects which have a method with the name `close` does not build on uniffi-rs `0.31`. I made a ticket here: https://github.com/mozilla/uniffi-rs/issues/2955. I have worked around this by renaming `close` -> `close_stream` as a kotlin specific override here: https://github.com/livekit/rust-sdks/pull/1286/commits/1fe0eba3225101e27f266ba5c2b2eced70e31a65

2. Any uniffi enums which have tagged cases that contain fields names `message` fail to build:
```
UniFFI emits, for a structured error variant carrying a `message` field:

        class AbnormalEnd(val `message`: kotlin.String) : DataStreamException() {
            override val message get() = "message=${ `message` }"
        }

    The constructor property and the overridden `Throwable.message` are both called
    `message` in one class body, which does not compile (and their types differ --
    String vs String? -- so they cannot be merged into one override either).

    The upstream fix would be for uniffi to avoid emitting a property that collides with
    Throwable.message (or for livekit-uniffi to not name the field `message`).
```
Unfortunately, this one can't be fixed in the same way, so I've opted to rename all error cases that contain `message` to instead be `reason`, also in here: https://github.com/livekit/rust-sdks/pull/1286/commits/1fe0eba3225101e27f266ba5c2b2eced70e31a65. The uniffi-rs maintainers seem to be unable to figure out a good way to fix this - the latest related issue was closed: https://github.com/mozilla/uniffi-rs/issues/2938

-----------------

> [!WARNING]
> This pull request was LLM generated and has only been reviewed by a human who isn't a domain expert in android development. I have tested this and confirms it works in the happy path, but no other validation has been done.
> 
> A more thorough review of this needs to occur before it could be merged.

## Todo
- [ ] Get https://github.com/livekit/rust-sdks/pull/1286 merged, the kotlin bindings published, and this ported over to use the kotlin bindings from maven central (instead of maven local)
- [ ] Add changeset

### CODE REVIEW

File: livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt [hunk 1]
```diff
@@ -0,0 +1,678 @@
...
+    /**
+     * Outbound packets waiting to go on the wire.
+     *
+     * The FFI delegate is a plain synchronous callback on a Rust runtime thread: it can neither
+     * block nor suspend, but sending has to await publisher connection and data channel
+     * backpressure. Handing off through an unbounded channel drained by a single coroutine keeps
+     * packets in the order the core emitted them while restoring the backpressure the previous
+     * implementation had.
+     */
+    private val outboundPackets = Channel<ByteArray>(Channel.UNLIMITED)
```
[the remark this answers is not in the document]
Comment (1egoman) [reply livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt:159]: I left a comment covering this in the swift version: https://github.com/livekit/client-sdk-swift/pull/975#discussion_r3722653709. IMO it's worth reading. Lukas + I (mostly lukas) ended up addressing this same issue on the web via a new "low water mark / high water mark" approach on the data channel which I think if android were to also adopt, it could fix this: https://github.com/livekit/client-sdk-js/pull/2014

IMO fixing this should be out of scope and be a follow up task though, as this pull request is already quite large and this would easily add many hundreds more lines.

File: livekit-android-sdk/src/main/java/io/livekit/android/room/Room.kt [hunk 2]
```diff
@@ -1341,22 +1362,10 @@ constructor(
...
-                incomingDataStreamManager.handleStreamTrailer(dp.streamTrailer, encryptionType)
-            }
-
-            // Ignore other cases.
-            else -> {}
-        }
+        // Forwarded whole rather than destructured: the core re-decodes the packet itself, and v2
+        // headers carry fields (inline content, compression) that only it interprets. The packet has
+        // already been decrypted by the engine, which is what the core expects.
+        dataStreams.handleIncoming(dp)
```
[the remark this answers is not in the document]
Comment (1egoman) [reply livekit-android-sdk/src/main/java/io/livekit/android/room/Room.kt]: I think this is a side effect of the rust core being decoupled from e2ee. Maybe it's worth re-adding this check back in though before the message gets sent into rust?

File: livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt [hunk 3]
```diff
@@ -0,0 +1,678 @@
...
+     */
+    private val ffiDispatcher: CoroutineDispatcher = Dispatchers.IO
+
+    private val coroutineScope = CoroutineScope(SupervisorJob() + ffiDispatcher)
+
+    private val textStreamHandlers = Collections.synchronizedMap(mutableMapOf<String, TextStreamHandler>())
+    private val byteStreamHandlers = Collections.synchronizedMap(mutableMapOf<String, ByteStreamHandler>())
+
+    /** Topics we have already warned about, so an unhandled topic logs once rather than per stream. */
+    private val warnedTopics = Collections.synchronizedSet(mutableSetOf<String>())
```
[the remark this answers is not in the document]
Comment (1egoman) [reply livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt]: This should be dropped, there should be a log per stream every time one comes in on an unknown topic (this matches the behavior in web and I believe swift). Must have missed this when I reviewed it, oops!

File: livekit-android-sdk/src/main/java/io/livekit/android/room/Room.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (1egoman) [reply livekit-android-sdk/src/main/java/io/livekit/android/room/Room.kt]: Related: https://github.com/livekit/client-sdk-android/pull/997#discussion_r3736947531. Yea, I think the check should be readded.

File: livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (1egoman) [reply livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt]: Fixed in b46b23c

[the remark this answers is not in the document]
Comment (1egoman) [reply livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt]: Fixed in b46b23c

File: livekit-android-sdk/src/main/java/io/livekit/android/room/Room.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (1egoman) [reply livekit-android-sdk/src/main/java/io/livekit/android/room/Room.kt]: Fixed in ec3ade1

[the remark this answers is not in the document]
Comment (1egoman) [reply livekit-android-sdk/src/main/java/io/livekit/android/room/Room.kt]: Fixed in ec3ade1

File: livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt [hunk 4]
```diff
@@ -213,21 +231,112 @@ internal constructor(
...
+            LivekitModels.DataPacket.ValueCase.STREAM_CHUNK ->
+                return matchesHeader(packet.streamChunk.streamId, encryptionType)
+
+            LivekitModels.DataPacket.ValueCase.STREAM_TRAILER ->
+                return matchesHeader(packet.streamTrailer.streamId, encryptionType)
+
+            else -> {}
+        }
+        return true
+    }
```
Comment (1egoman) [inline livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt:285]: Admittedly, I'm not 100% sure this is correct - I would think the final statement in this function should be `return false` since packets which are not end to end encrypted should be dropped?. According to a LLM it matches what a few other sdks do currently but I think this needs a second pair of eyes to confirm more familiar with how e2ee works.

File: livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt [hunk 5]
```diff
@@ -0,0 +1,809 @@
...
+            return
+        }
+
+        engine.waitForBufferStatusLow(packet.kind)
+        val result = engine.sendData(packet)
+        if (result.isFailure) {
+            // The core acknowledges the send as soon as it hands the packet over, so there is
+            // nobody left to return this to; the originating send call has already returned.
+            LKLog.w(result.exceptionOrNull()) { "Failed to send a data stream packet." }
+        }
```
[the remark this answers is not in the document]
Comment (1egoman) [reply livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt:406]: I will leave this up to @davidliu, this code isn't called as part of the call stack of `sendText` / etc - it's forwarded along from the `outboundPackets` channel:

```kotlin
        coroutineScope.launch {
            for (packet in outboundPackets) {
                sendPacket(packet)
            }
        }
```

This is IMO a fairly nice decoupling but not having access to the error is sort of the natural consequence of doing this. Maybe it's worth passing some sort of [oneshot channel](https://doc.rust-lang.org/std/sync/oneshot/fn.channel.html) / callback type interface where the caller can wait for `sendPacket` to signal back to the caller a success / error state? I'm unsure and would prefer to defer to DL on this.

File: settings.gradle [hunk 6]
```diff
@@ -17,6 +17,7 @@ dependencyResolutionManagement {
         google()
         mavenCentral()
         maven { url 'https://jitpack.io' }
+        mavenLocal()
```
[the remark this answers is not in the document]
Comment (1egoman) [reply settings.gradle:20]: This is marked as a TODO in the pull request body.

File: livekit-android-sdk/src/main/java/io/livekit/android/ConnectOptions.kt [hunk 7]
```diff
@@ -60,7 +60,7 @@ data class ConnectOptions(
      * for peer-to-peer feature negotiation (RPC v2, etc.). Defaults to the latest
      * version supported by this SDK build.
      */
-    val clientProtocol: ClientProtocolVersion = ClientProtocolVersion.DATA_STREAM_RPC,
+    val clientProtocol: ClientProtocolVersion = ClientProtocolVersion.DATA_STREAM_V2,
```
[the remark this answers is not in the document]
Comment (1egoman) [reply livekit-android-sdk/src/main/java/io/livekit/android/ConnectOptions.kt:63]: This is marked as a TODO in the pull request body.

File: livekit-android-test/src/main/java/io/livekit/android/test/util/AwaitCondition.kt [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    message: String = "Condition was not met",
+    condition: () -> Boolean,
+) {
+    val deadline = System.currentTimeMillis() + timeoutMs
+    while (System.currentTimeMillis() < deadline) {
+        scheduler.runCurrent()
+        if (condition()) {
+            return
+        }
+        Thread.sleep(POLL_MS)
```
Comment (xianshijing-lk) [inline livekit-android-test/src/main/java/io/livekit/android/test/util/AwaitCondition.kt:59]: I wonder if  Thread.yield() is better than Thread.sleep(POLL_MS) here ?

File: livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/StreamOptions.kt [hunk 9]
```diff
@@ -51,4 +65,10 @@ data class StreamBytesOptions(
      * The total exact size in bytes, if known.
      */
     val totalSize: Long? = null,
+    /**
+     * Whether the payload may be compressed, when every recipient supports it.
+     *
+     * @see StreamTextOptions.compress
+     */
+    val compress: Boolean = true,
```
Comment (xianshijing-lk) [inline livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/StreamOptions.kt:73]: curiously, should default compress to false ?

Comment (1egoman) [reply livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/StreamOptions.kt:73]: On other platforms, compress defaults to true, so I think keeping this true is the right thing to do for consistency.

File: livekit-android-sdk/src/main/java/io/livekit/android/room/ClientCapability.kt [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+ * The capabilities this SDK advertises to the server and, through it, to peers.
+ *
+ * Declared once so that every place which announces capabilities agrees. Data streams v2
+ * compression is advertised unconditionally: deflate-raw is compressed and decompressed by the
+ * Rust data stream core, so support does not vary by device, API level, or room options.
+ *
+ * @suppress
+ */
+internal val ADVERTISED_CLIENT_CAPABILITIES = listOf(
+    ClientCapability.COMPRESSION_DEFLATE_RAW,
```
Comment (MaxHeimbrock) [inline livekit-android-sdk/src/main/java/io/livekit/android/room/ClientCapability.kt:82]: Why don't we also advertise the `PACKET_TRAILER` capability for our Android client?

**Comment by MaxHeimbrock** [comment]
Nice job, I tested broadcasting data streams between v1 and v2 clients and it all was sent and received correctly. The room contained two v2 clients and one v1 client.

File: livekit-android-sdk/src/main/java/io/livekit/android/room/ClientCapability.kt (hunk 10 above)
Comment (1egoman) [reply livekit-android-sdk/src/main/java/io/livekit/android/room/ClientCapability.kt:82]: AFAIK, android does not support stripping / processing packet trailers from video frames. If this capability isn't advertised, I believe the SFU will do the trailer stripping instead, but I may be misremembering, cc @chenosaurus for clarification here.

Once the android client is updated to support packet trailer processing (ie, stripping them out, and also extracting the metadata and providing apis to consume this metadata downstream) then this capability can be added to the advertisement here.

File: livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (1egoman) [reply livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt]: This has changed a fair bit now that https://github.com/livekit/client-sdk-android/pull/997/commits/784240a86bded18a0da89ae9884e74a13094564c is committed.

I've moved away from the channel approach and instead now that `onPacketsAvailable` is a `suspend fun`, it doesn't have to be as necessarily decoupled.

File: livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt (hunk 4 above)
Comment (1egoman) [reply livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt]: Now with 70e4a75, this has also changed - all this code is moved into the rust core.

**Comment by 1egoman** [comment]
@MaxHeimbrock and/or @davidliu - FYI that Blaze [reviewed my rust change](https://github.com/livekit/rust-sdks/pull/1286#pullrequestreview-4951982568) and surfaced a few comments when integrating this into swift. These changes [led to some rust sdk FFI changes](https://github.com/livekit/rust-sdks/pull/1286/changes/e3c6be50f50c6dddf02836452c904a6ee5f6b362..1da01b49c4dc26a0dbb4690e4bbcb1d2935df653) which I have ported this PR to take into account as well. IMO, it would be worth another test here to make sure that everything still works, with a focus on the points blaze layed out in his review.

File: livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt [hunk 11]
```diff
@@ -0,0 +1,714 @@
...
+        coroutineScope.cancel()
+        // Releases the native handles, and with them the core's reference to our delegates. Those
+        // delegates are held by a static handle map on the way in, so skipping this would keep this
+        // object -- and through it the engine -- reachable for the life of the process.
+        synchronized(incomingLock) {
+            incoming?.destroy()
+            incoming = null
+        }
+        outgoing.destroy()
+    }
```
[the remark this answers is not in the document]
Comment (1egoman) [reply livekit-android-sdk/src/main/java/io/livekit/android/room/datastream/DataStreams.kt:342]: Fixed in bcc91bb
